go/vt/mysqlctl/schema.go

```diff
@@ -533,6 +533,52 @@ func (mysqld *Mysqld) ApplySchemaChange(ctx context.Context, dbName string, chan
 	return &tabletmanagerdatapb.SchemaChangeResult{BeforeSchema: beforeSchema, AfterSchema: afterSchema}, nil
 }
 
+// GetPrimaryKeyEquivalentColumns can be used if the table has
+// no defined PRIMARY KEY. It will return the columns in a
+// viable PRIMARY KEY equivalent (PKE) -- a NON-NULL UNIQUE
+// KEY -- in the specified table. When multiple PKE indexes
+// are available it will use the one with the fewest columns
+// in it.
+// If this function is used on a table that DOES have a
+// defined PRIMARY KEY then it may return the columns for
+// that index if it has the fewest columns in it among the
+// available PKE indexes on the table.
```

Comment:
Can we also optimize for `INT` columns over textual columns?

Reply:
Nice! Thank you for pointing out that query. I had not seen it but I did just read about that behavior here a few days ago: https://vitess.io/docs/14.0/design-docs/vreplication/keys/

Any reason _NOT_ to re-use the same query that's being used for OnlineDDL? I like that added optimization of favoring INTEGER types in ASC capacity/size over CHAR/BYTE based types. That would also make the documentation relatively simple and more unified: *generalize the noted docs page above a bit more as it would apply to VReplication more generally*.

I'm curious if we would still require these fields and the code around them for OnlineDDL:
```
    source_unique_key_columns:"order_id,customer_id" 
    target_unique_key_columns:"order_id,cust_id" 
    source_unique_key_target_columns:"order_id,cust_id"
```

I'm guessing there's still some cases for it but wanted to ask. 

Thank you for looking at this and sharing the info! ❤️

Reply:
The query is still heuristic. It only inspects the data type of the 1st column, and does not take into consideration the "average" data type for all columns.
Obviously we want to only use non-nullable keys (the query includes nullable keys, but clearly indicates them so it's easy to filter those.

@mattlord I don't know if you've seen [this](https://github.com/vitessio/vitess/pull/8364) PR to support non-PK keys in `vreplication`. This was a year ago and I need a refresher, but basically we can write the names of an "overriding" unqiue key columns into the vreplication's `Rule`:

https://github.com/vitessio/vitess/pull/8364/files#diff-5e69c397e15dcfda72bb6fe8dba93b2d11a8d2d7b594ebfbbdcfdc07bfc3791d

Reply:
Yeah, I think it's worth leaving that in as having an override can certainly be useful. 

I've taken the index column types into account now in this query. 🙇